File: TheWatch-Android/app/src/main/java/com/thewatch/app/data/sync/SyncEngine.kt
/**
 * +----------------------------------------------------------------------+
 * | WRITE-AHEAD LOG                                                       |
 * +----------------------------------------------------------------------+
 * | File:         SyncEngine.kt                                           |
 * | Purpose:      Singleton coordinator for all offline-first sync        |
 * |               operations across TheWatch. Manages a priority queue    |
 * |               of SyncTask items persisted in Room, dispatches them    |
 * |               through SyncDispatcher when connectivity is available,  |
 * |               and handles task coalescing for duplicate entity updates.|
 * | Created:      2026-03-24                                              |
 * | Author:       Claude                                                  |
 * | Dependencies: SyncTaskDao, SyncDispatcher, ConnectivityMonitor,       |
 * |               WatchLogger                                             |
 * |                                                                       |
 * | Usage example:                                                        |
 * |   // Enqueue from anywhere in the app:                                |
 * |   syncEngine.enqueue(                                                 |
 * |       entityType = SyncEntityType.SOS_EVENT,                          |
 * |       entityId = "sos-abc",                                           |
 * |       action = SyncTaskAction.CREATE,                                 |
 * |       payload = sosJson,                                              |
 * |       priority = SyncPriority.CRITICAL                                |
 * |   )                                                                   |
 * |                                                                       |
 * |   // The engine will:                                                 |
 * |   // 1. Persist to Room immediately (survives process death)          |
 * |   // 2. If online, trigger immediate flush via SyncWorker             |
 * |   // 3. If offline, ConnectivityMonitor triggers flush on reconnect   |
 * |                                                                       |
 * | Coalescing: If an UPDATE for the same entityType+entityId is already  |
 * | QUEUED, the engine replaces the payload (last-write-wins locally)     |
 * | rather than creating a duplicate task.                                |
 * |                                                                       |
 * | NOTE: The engine does NOT hold tasks in memory. Room is the single    |
 * | source of truth. This ensures no data loss on process death.          |
 * | Consider adding metrics emission (enqueue rate, flush latency,        |
 * | dead-letter rate) for operational dashboards.                         |
 * +----------------------------------------------------------------------+
 */
package com.thewatch.app.data.sync

import android.content.Context
import android.util.Log
import com.thewatch.app.data.logging.WatchLogger
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch
import java.util.UUID
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.min
import kotlin.math.pow

/**
 * Callback interface for sync providers that want to register with the engine.
 * The existing logging sync (OfflineSyncWorker/NativeLogSyncAdapter) registers
 * through this interface so it participates in the generalized engine without
 * breaking its existing behavior.
 */
interface SyncProvider {
    /** The entity type this provider handles. */
    val entityType: SyncEntityType

    /**
     * Called by the engine when a task for this provider's entity type is ready
     * to be dispatched. Providers may perform custom serialization, validation,
     * or side effects before the standard Firestore push.
     *
     * Return true to let the engine proceed with standard dispatch.
     * Return false if the provider handled the push itself.
     */
    suspend fun onBeforeDispatch(task: SyncTaskEntity): Boolean

    /**
     * Called after successful dispatch. Providers can update local state
     * (e.g., mark a log entry as synced in the logging subsystem's own table).
     */
    suspend fun onAfterDispatch(task: SyncTaskEntity, serverId: String?)
}

@Singleton
class SyncEngine @Inject constructor(
    @ApplicationContext private val context: Context,
    private val syncTaskDao: SyncTaskDao,
    private val syncDispatcher: SyncDispatcher,
    private val connectivityMonitor: ConnectivityMonitor,
    private val logger: WatchLogger
) {
    companion object {
        private const val TAG = "TheWatch.SyncEngine"
        private const val MAX_RETRIES_DEFAULT = 5
        private const val BASE_BACKOFF_SECONDS = 30L
        private const val MAX_BACKOFF_SECONDS = 900L // 15 min
        private const val BATCH_SIZE = 50
        private const val PRUNE_COMPLETED_HOURS = 24L
        private const val PRUNE_DEAD_LETTER_DAYS = 7L
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    /** Registered sync providers, keyed by entity type. */
    private val providers = mutableMapOf<SyncEntityType, SyncProvider>()

    /**
     * Register a sync provider for a specific entity type.
     * This allows existing subsystems (logging, evidence, etc.) to plug
     * into the generalized engine with custom pre/post-dispatch hooks.
     */
    fun registerProvider(provider: SyncProvider) {
        providers[provider.entityType] = provider
        Log.i(TAG, "Registered sync provider for ${provider.entityType}")
    }

    /**
     * Enqueue a sync task. Persists immediately to Room.
     * If online and priority is CRITICAL or HIGH, triggers an expedited flush.
     *
     * @param entityType What kind of data
     * @param entityId Unique ID of the entity
     * @param action Create/Update/Delete
     * @param payload JSON-serialized data
     * @param priority See [SyncPriority]
     * @param userId Owner of this data
     * @param idempotencyKey Custom dedup key, defaults to task ID
     * @return The generated task ID
     */
    suspend fun enqueue(
        entityType: SyncEntityType,
        entityId: String,
        action: SyncTaskAction,
        payload: String,
        priority: Int = SyncPriority.NORMAL,
        userId: String = "",
        idempotencyKey: String? = null
    ): String {
        // Coalesce: if an UPDATE for the same entity is already queued, replace payload
        if (action == SyncTaskAction.UPDATE) {
            val existing = syncTaskDao.findPendingForEntity(entityType, entityId)
            if (existing != null && existing.action == SyncTaskAction.UPDATE) {
                val updated = existing.copy(
                    payload = payload,
                    priority = minOf(existing.priority, priority), // Escalate priority
                    createdAt = System.currentTimeMillis()
                )
                syncTaskDao.update(updated)
                Log.d(TAG, "Coalesced UPDATE for $entityType/$entityId (task=${existing.id})")
                triggerFlushIfNeeded(priority)
                return existing.id
            }
        }

        val taskId = UUID.randomUUID().toString()
        val task = SyncTaskEntity(
            id = taskId,
            entityType = entityType,
            entityId = entityId,
            action = action,
            payload = payload,
            priority = priority,
            userId = userId,
            idempotencyKey = idempotencyKey ?: taskId
        )

        syncTaskDao.insert(task)
        Log.i(TAG, "Enqueued $action for $entityType/$entityId (priority=$priority, id=$taskId)")

        triggerFlushIfNeeded(priority)
        return taskId
    }

    /**
     * Convenience: enqueue a batch of tasks atomically.
     */
    suspend fun enqueueBatch(tasks: List<SyncTaskEntity>) {
        syncTaskDao.insertAll(tasks)
        Log.i(TAG, "Enqueued batch of ${tasks.size} tasks")
        val highestPriority = tasks.minOfOrNull { it.priority } ?: SyncPriority.NORMAL
        triggerFlushIfNeeded(highestPriority)
    }

    /**
     * Main flush cycle. Called by SyncWorker on schedule and on connectivity restored.
     * Processes all pending tasks in priority order, dispatching through SyncDispatcher.
     *
     * @return Pair of (successCount, failCount)
     */
    suspend fun flush(): Pair<Int, Int> {
        Log.i(TAG, "flush() started")

        // 1. Recover stuck tasks from a previous crash
        syncTaskDao.resetStuckTasks()

        // 2. Check backend reachability
        if (!syncDispatcher.isBackendReachable()) {
            Log.w(TAG, "Backend not reachable — skipping flush")
            return Pair(0, 0)
        }

        // 3. Fetch all pending tasks ordered by priority
        val tasks = syncTaskDao.getPendingByPriority(MAX_RETRIES_DEFAULT)
        if (tasks.isEmpty()) {
            Log.i(TAG, "No pending tasks")
            return Pair(0, 0)
        }

        Log.i(TAG, "Processing ${tasks.size} pending tasks")

        var successCount = 0
        var failCount = 0

        // 4. Process in batches
        for (batch in tasks.chunked(BATCH_SIZE)) {
            for (task in batch) {
                // Per-task exponential backoff
                if (shouldDefer(task)) {
                    Log.d(TAG, "Deferring task ${task.id} (retry ${task.retryCount})")
                    continue
                }

                syncTaskDao.markInProgress(task.id)

                // Let registered provider run pre-dispatch hook
                val provider = providers[task.entityType]
                val shouldContinueDispatch = try {
                    provider?.onBeforeDispatch(task) ?: true
                } catch (e: Exception) {
                    Log.w(TAG, "Provider pre-dispatch failed for ${task.id}", e)
                    true // Continue with standard dispatch on provider error
                }

                if (!shouldContinueDispatch) {
                    // Provider handled the push itself
                    syncTaskDao.markCompleted(task.id)
                    successCount++
                    continue
                }

                // Dispatch through the generalized dispatcher
                when (val result = syncDispatcher.dispatch(task)) {
                    is SyncDispatchResult.Success -> {
                        syncTaskDao.markCompleted(task.id)
                        successCount++

                        // Notify provider of success
                        try {
                            provider?.onAfterDispatch(task, result.serverId)
                        } catch (e: Exception) {
                            Log.w(TAG, "Provider post-dispatch failed for ${task.id}", e)
                        }

                        Log.d(TAG, "Synced ${task.entityType}/${task.entityId} -> ${result.serverId}")
                    }

                    is SyncDispatchResult.RetryableFailure -> {
                        val newRetryCount = task.retryCount + 1
                        if (newRetryCount >= task.maxRetries) {
                            syncTaskDao.markDeadLetter(task.id)
                            Log.e(TAG, "Task ${task.id} moved to dead-letter after $newRetryCount retries: ${result.message}")
                        } else {
                            syncTaskDao.markFailed(task.id, result.message, newRetryCount)
                            Log.w(TAG, "Retryable failure for ${task.id}: ${result.message}")
                        }
                        failCount++

                        // If dispatcher signals circuit open, stop batch
                        if (syncDispatcher.isCircuitOpen()) {
                            Log.w(TAG, "Circuit breaker open — stopping flush")
                            return Pair(successCount, failCount)
                        }
                    }

                    is SyncDispatchResult.PermanentFailure -> {
                        syncTaskDao.markDeadLetter(task.id)
                        failCount++
                        Log.e(TAG, "Permanent failure for ${task.id}: ${result.message}")
                    }
                }
            }
        }

        // 5. Prune old completed and dead-letter tasks
        val completedCutoff = System.currentTimeMillis() - TimeUnit.HOURS.toMillis(PRUNE_COMPLETED_HOURS)
        syncTaskDao.pruneCompleted(completedCutoff)

        val deadLetterCutoff = System.currentTimeMillis() - TimeUnit.DAYS.toMillis(PRUNE_DEAD_LETTER_DAYS)
        syncTaskDao.pruneDeadLetter(deadLetterCutoff)

        Log.i(TAG, "flush() complete — synced=$successCount, failed=$failCount")
        return Pair(successCount, failCount)
    }

    /**
     * Flush only tasks of a specific entity type.
     * Used for targeted sync (e.g., "sync all pending SOS events NOW").
     */
    suspend fun flushEntityType(entityType: SyncEntityType): Pair<Int, Int> {
        syncTaskDao.resetStuckTasks()

        val tasks = syncTaskDao.getPendingByEntityType(entityType, MAX_RETRIES_DEFAULT)
        if (tasks.isEmpty()) return Pair(0, 0)

        var success = 0
        var fail = 0

        for (task in tasks) {
            syncTaskDao.markInProgress(task.id)
            when (val result = syncDispatcher.dispatch(task)) {
                is SyncDispatchResult.Success -> {
                    syncTaskDao.markCompleted(task.id)
                    success++
                    providers[task.entityType]?.onAfterDispatch(task, result.serverId)
                }
                is SyncDispatchResult.RetryableFailure -> {
                    val newRetry = task.retryCount + 1
                    if (newRetry >= task.maxRetries) {
                        syncTaskDao.markDeadLetter(task.id)
                    } else {
                        syncTaskDao.markFailed(task.id, result.message, newRetry)
                    }
                    fail++
                }
                is SyncDispatchResult.PermanentFailure -> {
                    syncTaskDao.markDeadLetter(task.id)
                    fail++
                }
            }
        }

        return Pair(success, fail)
    }

    /**
     * Re-queue a dead-letter task for manual retry.
     */
    suspend fun retryDeadLetter(taskId: String) {
        syncTaskDao.requeue(taskId)
        Log.i(TAG, "Re-queued dead-letter task $taskId")
        triggerFlushIfNeeded(SyncPriority.HIGH)
    }

    /**
     * Observe the number of pending tasks (for UI badge).
     */
    fun observePendingCount(): Flow<Int> = syncTaskDao.observePendingCount()

    /**
     * Clear all pending sync tasks (used on logout).
     */
    suspend fun clearAll() {
        syncTaskDao.clearAll()
        Log.i(TAG, "All sync tasks cleared")
    }

    // ── Private Helpers ──────────────────────────────────────────────

    /**
     * Trigger an expedited WorkManager sync if the task is high priority
     * and we're currently online.
     */
    private fun triggerFlushIfNeeded(priority: Int) {
        if (priority <= SyncPriority.HIGH && connectivityMonitor.isOnline()) {
            SyncWorker.enqueueExpedited(context)
        }
    }

    /**
     * Per-task exponential backoff: min(BASE * 2^retryCount, MAX).
     */
    private fun shouldDefer(task: SyncTaskEntity): Boolean {
        if (task.retryCount == 0 || task.lastAttemptAt == null) return false

        val backoffSeconds = min(
            BASE_BACKOFF_SECONDS * 2.0.pow(task.retryCount.toDouble()).toLong(),
            MAX_BACKOFF_SECONDS
        )
        val nextAllowedTime = task.lastAttemptAt + (backoffSeconds * 1000)
        return System.currentTimeMillis() < nextAllowedTime
    }
}

File: TheWatch-iOS/Services/Sync/SyncEngine.swift
/**
 * ┌──────────────────────────────────────────────────────────────────────┐
 * │ WRITE-AHEAD LOG                                                      │
 * ├──────────────────────────────────────────────────────────────────────┤
 * │ File:         SyncEngine.swift                                       │
 * │ Purpose:      Singleton coordinator for all offline-first sync       │
 * │               operations on iOS. Manages a priority queue persisted  │
 * │               via SyncTaskStore, dispatches through SyncDispatcher,  │
 * │               and handles task coalescing for duplicate updates.     │
 * │               Mirrors Android's SyncEngine.kt.                       │
 * │ Created:      2026-03-24                                             │
 * │ Author:       Claude                                                 │
 * │ Dependencies: SyncTaskStore, SyncDispatcher, SyncConnectivityMonitor│
 * │                                                                      │
 * │ Usage example:                                                       │
 * │   // Enqueue from anywhere:                                          │
 * │   await SyncEngine.shared.enqueue(                                   │
 * │       entityType: .sosEvent,                                         │
 * │       entityId: "sos-abc",                                           │
 * │       action: .create,                                               │
 * │       payload: sosJSON,                                              │
 * │       priority: .critical                                            │
 * │   )                                                                  │
 * │                                                                      │
 * │   // Engine persists to disk immediately (survives process death).   │
 * │   // If online, triggers immediate flush.                            │
 * │   // If offline, ConnectivityMonitor triggers flush on reconnect.    │
 * │                                                                      │
 * │ Provider registration:                                               │
 * │   SyncEngine.shared.registerProvider(logSyncProvider)                │
 * │   // Existing logging sync can plug in without breaking.             │
 * │                                                                      │
 * │ NOTE: The engine does NOT hold tasks in memory — SyncTaskStore       │
 * │ (file-backed actor) is the single source of truth.                   │
 * └──────────────────────────────────────────────────────────────────────┘
 */

import Foundation
import Combine
import os.log

/// Protocol for sync providers that register with the engine.
/// Mirrors Android's SyncProvider interface.
protocol SyncProviderProtocol {
    var entityType: SyncEntityType { get }

    /// Called before dispatch. Return true to let engine proceed, false if provider handled it.
    func onBeforeDispatch(_ task: SyncTask) async -> Bool

    /// Called after successful dispatch for post-sync bookkeeping.
    func onAfterDispatch(_ task: SyncTask, serverId: String?) async
}

@Observable
final class SyncEngine {

    // MARK: - Singleton

    static let shared = SyncEngine()

    // MARK: - Dependencies

    let store = SyncTaskStore()
    let dispatcher = SyncDispatcher()
    let connectivityMonitor = SyncConnectivityMonitor()

    // MARK: - Config

    private let maxRetries = 5
    private let baseBackoffSeconds: TimeInterval = 30
    private let maxBackoffSeconds: TimeInterval = 900
    private let batchSize = 50
    private let pruneCompletedHours: TimeInterval = 24 * 3600
    private let pruneDeadLetterDays: TimeInterval = 7 * 24 * 3600

    // MARK: - State

    private(set) var pendingCount: Int = 0
    private var providers: [SyncEntityType: SyncProviderProtocol] = [:]
    private var cancellables = Set<AnyCancellable>()
    private let logger = Logger(subsystem: "com.thewatch.app", category: "SyncEngine")

    // MARK: - Init

    private init() {
        // Wire connectivity monitor to trigger flush on reconnect
        connectivityMonitor.onReconnect = { [weak self] in
            Task {
                await self?.flush()
            }
        }

        // Observe pending count
        store.pendingCountPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] count in
                self?.pendingCount = count
            }
            .store(in: &cancellables)
    }

    // MARK: - Provider Registration

    func registerProvider(_ provider: SyncProviderProtocol) {
        providers[provider.entityType] = provider
        logger.info("Registered sync provider for \(provider.entityType.rawValue)")
    }

    // MARK: - Enqueue

    /// Enqueue a sync task. Persists immediately. Triggers expedited flush for high-priority tasks.
    @discardableResult
    func enqueue(
        entityType: SyncEntityType,
        entityId: String,
        action: SyncTaskAction,
        payload: String,
        priority: SyncPriority = .normal,
        userId: String = "",
        idempotencyKey: String? = nil
    ) async -> String {

        // Coalesce: if UPDATE for same entity is already queued, replace payload
        if action == .update {
            if let existing = await store.findPendingForEntity(entityType: entityType, entityId: entityId),
               existing.action == .update {
                var updated = existing
                updated.payload = payload
                updated.priority = min(existing.priority, priority)
                updated.createdAt = Date()
                try? await store.update(updated)
                logger.debug("Coalesced UPDATE for \(entityType.rawValue)/\(entityId)")
                triggerFlushIfNeeded(priority: priority)
                return existing.id
            }
        }

        let task = SyncTask(
            entityType: entityType,
            entityId: entityId,
            action: action,
            payload: payload,
            priority: priority,
            userId: userId,
            idempotencyKey: idempotencyKey
        )

        try? await store.insert(task)
        logger.info("Enqueued \(action.rawValue) for \(entityType.rawValue)/\(entityId) (priority=\(priority.rawValue))")

        triggerFlushIfNeeded(priority: priority)
        return task.id
    }

    /// Enqueue a batch of tasks.
    func enqueueBatch(_ tasks: [SyncTask]) async {
        try? await store.insertAll(tasks)
        logger.info("Enqueued batch of \(tasks.count) tasks")
        if let highestPriority = tasks.min(by: { $0.priority < $1.priority })?.priority {
            triggerFlushIfNeeded(priority: highestPriority)
        }
    }

    // MARK: - Flush

    /// Main flush cycle. Called by SyncWorker (BGProcessingTask) and on connectivity restored.
    @discardableResult
    func flush() async -> (success: Int, failed: Int) {
        logger.info("flush() started")

        // 1. Recover stuck tasks
        try? await store.resetStuckTasks()

        // 2. Check backend
        guard await dispatcher.isBackendReachable() else {
            logger.warning("Backend not reachable — skipping flush")
            return (0, 0)
        }

        // 3. Get pending tasks
        let tasks = await store.getPendingByPriority(maxRetries: maxRetries)
        guard !tasks.isEmpty else {
            logger.info("No pending tasks")
            return (0, 0)
        }

        logger.info("Processing \(tasks.count) pending tasks")

        var successCount = 0
        var failCount = 0

        // 4. Process in batches
        for batch in tasks.chunked(into: batchSize) {
            for task in batch {
                // Per-task exponential backoff
                if shouldDefer(task) {
                    logger.debug("Deferring task \(task.id) (retry \(task.retryCount))")
                    continue
                }

                try? await store.markInProgress(task.id)

                // Provider pre-dispatch hook
                let provider = providers[task.entityType]
                let shouldContinue = await (provider?.onBeforeDispatch(task) ?? true)

                if !shouldContinue {
                    try? await store.markCompleted(task.id)
                    successCount += 1
                    continue
                }

                // Dispatch
                let result = await dispatcher.dispatch(task)

                switch result {
                case .success(let serverId):
                    try? await store.markCompleted(task.id)
                    successCount += 1
                    await provider?.onAfterDispatch(task, serverId: serverId)
                    logger.debug("Synced \(task.entityType.rawValue)/\(task.entityId) -> \(serverId ?? "nil")")

                case .retryableFailure(let message, _):
                    let newRetryCount = task.retryCount + 1
                    if newRetryCount >= task.maxRetries {
                        try? await store.markDeadLetter(task.id)
                        logger.error("Task \(task.id) moved to dead-letter: \(message)")
                    } else {
                        try? await store.markFailed(task.id, error: message, retryCount: newRetryCount)
                        logger.warning("Retryable failure for \(task.id): \(message)")
                    }
                    failCount += 1

                    if dispatcher.isCircuitOpen() {
                        logger.warning("Circuit breaker open — stopping flush")
                        return (successCount, failCount)
                    }

                case .permanentFailure(let message, _):
                    try? await store.markDeadLetter(task.id)
                    failCount += 1
                    logger.error("Permanent failure for \(task.id): \(message)")
                }
            }
        }

        // 5. Prune old tasks
        let completedCutoff = Date().addingTimeInterval(-pruneCompletedHours)
        try? await store.pruneCompleted(olderThan: completedCutoff)

        let deadLetterCutoff = Date().addingTimeInterval(-pruneDeadLetterDays)
        try? await store.pruneDeadLetter(olderThan: deadLetterCutoff)

        logger.info("flush() complete — synced=\(successCount), failed=\(failCount)")
        return (successCount, failCount)
    }

    /// Flush only tasks of a specific entity type.
    @discardableResult
    func flushEntityType(_ entityType: SyncEntityType) async -> (success: Int, failed: Int) {
        try? await store.resetStuckTasks()
        let tasks = await store.getPendingByEntityType(entityType, maxRetries: maxRetries)
        guard !tasks.isEmpty else { return (0, 0) }

        var success = 0
        var fail = 0

        for task in tasks {
            try? await store.markInProgress(task.id)
            let result = await dispatcher.dispatch(task)

            switch result {
            case .success(let serverId):
                try? await store.markCompleted(task.id)
                success += 1
                await providers[task.entityType]?.onAfterDispatch(task, serverId: serverId)
            case .retryableFailure(let msg, _):
                let newRetry = task.retryCount + 1
                if newRetry >= task.maxRetries {
                    try? await store.markDeadLetter(task.id)
                } else {
                    try? await store.markFailed(task.id, error: msg, retryCount: newRetry)
                }
                fail += 1
            case .permanentFailure:
                try? await store.markDeadLetter(task.id)
                fail += 1
            }
        }

        return (success, fail)
    }

    /// Re-queue a dead-letter task.
    func retryDeadLetter(_ taskId: String) async {
        try? await store.requeue(taskId)
        logger.info("Re-queued dead-letter task \(taskId)")
        triggerFlushIfNeeded(priority: .high)
    }

    /// Clear all tasks (on logout).
    func clearAll() async {
        try? await store.clearAll()
        logger.info("All sync tasks cleared")
    }

    // MARK: - Start / Stop

    func start() {
        connectivityMonitor.start()
        SyncBackgroundWorker.registerBackgroundTask()
        SyncBackgroundWorker.scheduleBackgroundSync()
        logger.info("SyncEngine started")
    }

    func stop() {
        connectivityMonitor.stop()
        logger.info("SyncEngine stopped")
    }

    // MARK: - Private

    private func triggerFlushIfNeeded(priority: SyncPriority) {
        if priority.rawValue <= SyncPriority.high.rawValue && connectivityMonitor.isConnected {
            Task { await flush() }
        }
    }

    private func shouldDefer(_ task: SyncTask) -> Bool {
        guard task.retryCount > 0, let lastAttempt = task.lastAttemptAt else { return false }

        let backoff = min(
            baseBackoffSeconds * pow(2.0, Double(task.retryCount)),
            maxBackoffSeconds
        )
        let nextAllowed = lastAttempt.addingTimeInterval(backoff)
        return Date() < nextAllowed
    }
}

// MARK: - Array Chunking

private extension Array {
    func chunked(into size: Int) -> [[Element]] {
        stride(from: 0, to: count, by: size).map {
            Array(self[$0 ..< Swift.min($0 + size, count)])
        }
    }
}
